[12.0][FIX] use the new methods for next period calculation result of the refactoring in #434 to consider the recurring_invoicing_type for monthlylastday mode and clean code

sbejaoui opened the pull request:
After the #434, the invoicing type (pre-paid and post-paid) became relevant for monthly last day mode. This PR fix the invoice date for the forecast of this recurring mode.

sbidoul commented:
@sbejaoui is this a fix or an improvement? If this is a fix, could you add, as first commit, a failing test that reveals the error, then as a second commit, the fix, then as subsequent commits other improvements? That will facilitate review a lot. Thanks!

sbejaoui commented:
> @sbejaoui is this a fix or an improvement? If this is a fix, could you add, as first commit, a failing test that reveals the error, then as a second commit, the fix, then as subsequent commits other improvements? That will facilitate review a lot. Thanks!

this PR fix the forecast generation for the monthlylastday to take in consideration the recurring_invoicing_type pre/post paid.

There is already two new unit tests to cover the fix in the first commit.

The second commit improve the code by using the new method for next period dates calculation.

sbejaoui committed:
[IMP] - add failing test for monthlylastday, pre-paid case

sbejaoui committed:
[FIX] - use the next_period dates to compute forecast dates

use the new methods for next period calculation result of the refactoring in #434
to consider the recurring_invoicing_type for monthlylastday mode and clean code

sbidoul reviewed (approved):
LGTM

sbidoul commented on contract_forecast/models/contract_line.py:
```diff
@@ -69,38 +69,41 @@ def _generate_forecast_periods(self):
                     )
-                    period_date_start, period_date_end, recurring_next_date = (
-                        period_dates
+                ):
```
I think this can be simplified by first initializing `recurring_next_date = rec.recurring_next_date; period_date_end = rec.next_period_date_end; period_date_start = rec.next_period_date_start` and then removing `if rec.recurring_next_date` and moving that test in the `while` condition. This is not blocking for this PR and can be done in a follow-up.

sbejaoui commented:
@gva-acsone, can you test and review this

gva-acsone reviewed (approved):
LGTM Functional test

sbejaoui commented:
/ocabot merge patch
